### PR TITLE
Add superpmi collect automation

### PR DESCRIPTION
**Pull request by jashook**
This commit includes scripting for collecting superPMI over the coreclr
tests and crossgen of the Framework Assemblies. It also includes the ability
to expand collection over any managed dll passed to the tool. It also includes
a script to upload, download and list the packages that are created by the tool
into Azure Blob Storage.

What is still to be done:

**Documentation**: see high level overview [here](https://github.com/dotnet/jitutils/issues/33)
- [SuperPMI Replay](https://github.com/dotnet/jitutils/issues/35): this adds the skeleton for replay, it still needs to be hooked up to download the mch files.
- [Expanding what we collect over](https://github.com/dotnet/jitutils/issues/34)
- [Automate by adding to Dotnet CI](https://github.com/dotnet/jitutils/issues/44)

### CODE REVIEW

**Comment by jashook** [comment]
See #32 for SuperPMI Automation work items.

**Comment by jashook** [comment]
@RussKeldorph @russellhadley @adiaaida please take a look when you can.

File: src/collect-all/Utilities.cs [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+      // buildType: Debug
+
+      if (string.IsNullOrEmpty(buildType))
+      {
+        buildType = "Debug";
+      }
+
+      if (string.IsNullOrEmpty(arch)) 
+      {
+         arch = RuntimeInformation.OSArchitecture.ToString();
```
Comment (adiaaida) [inline src/collect-all/Utilities.cs]: Why not just make this RuntimeInformation.OSArchitecture.ToString().ToLower() and then check against arch != "x64" to throw the exception?

File: src/collect-all/Utilities.cs [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+      this.m_testRoot = Path.Combine(coreclrRepoLocation, "bin", "tests", this.m_os + "." + this.m_arch + "." + this.m_buildType);
+      
+      string coreRootDirName = this.m_os == "Windows_NT" ? "Core_Root" : "coreoverlay";
+      this.m_coreRoot = Path.Combine(this.m_binLocation, "tests", this.m_os + "." + this.m_arch + "." + this.m_buildType, "Tests", coreRootDirName);
+
+      if (!Directory.Exists(this.m_testRoot))
+      {
+         if (this.m_os == "Windows_NT")
+         {
+           Console.WriteLine("Error, Tests were not built. Please build CoreCLR");
```
Comment (adiaaida) [inline src/collect-all/Utilities.cs]: Is this an error, or just a warning? If an error, shouldn't we throw?

File: src/collect-all/Utilities.cs [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+            job_name = "ubuntu14.04_release";
+         }
+         
+         string url = base_url + job_name + "/lastSuccessfulBuild/artifact/bin/build.tar.gz";
+         string args = $"{url} -P {corefxDirectory}"; 
+         
+         // Download the tar.gz
+         Process proc = new Process();
+      
+         proc.StartInfo.FileName = "wget";
```
Comment (adiaaida) [inline src/collect-all/Utilities.cs]: Should probably have this in a try/catch

File: src/collect-all/Utilities.cs [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+         {
+            throw new FileNotFoundException($"{buildGzLocation} not found.");
+         }
+         
+         args = $"xzf {corefxDirectory}/build.tar.gz -C {corefxDirectory}";
+         
+         // untar, unzip
+         proc = new Process();
+      
+         proc.StartInfo.FileName = "tar";
```
Comment (adiaaida) [inline src/collect-all/Utilities.cs]: Should probably have this in a try/catch

File: src/collect-all/Utilities.cs [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+      DirectoryInfo dir = new DirectoryInfo(source);
+      
+      if (!dir.Exists)
+      {
+         string err_msg = $"Source directory {source} does not exit";
+         throw new DirectoryNotFoundException(err_msg);
+      }
+      
+      // If the directory does not exist, create it.
+      if (!Directory.Exists(dest))
```
Comment (adiaaida) [inline src/collect-all/Utilities.cs]: You don't need to check if the dir already exists. If it does, CreateDirectory does nothing.

File: src/collect-all/Utilities.cs [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+   public void DownloadTests()
+   {
+      Directory.CreateDirectory(this.m_testRoot);
+
+      // Find the last successful test build based on OS.
+      string url = "http://dotnet-ci.cloudapp.net/job/dotnet_coreclr/job/master/job/release_windows_nt/lastSuccessfulBuild/artifact/bin/tests/tests.zip";
+      string args = $"{url} -P {this.m_testRoot}";
+         
+      // Download the zip
+      Process proc = new Process();
```
Comment (adiaaida) [inline src/collect-all/Utilities.cs]: try/catch

File: src/collect-all/Utilities.cs (hunk 2 above)
Comment (adiaaida) [inline src/collect-all/Utilities.cs]: As written, on Windows, we will still call DownloadTests(), which runs wget, but we can't assume wget can be found on Windows machines, so this should probably be an exception

File: src/collect-all/Utilities.cs [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+      
+      proc.StartInfo.FileName = "wget";
```
Comment (adiaaida) [inline src/collect-all/Utilities.cs]: Can't assume wget on windows

File: src/collect-all/Utilities.cs [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+   private class Setting
+   {
+      public string CoreClrRepoLocation;
+      public string BuildType;
+      public string Arch;
+   }
+   
+   public bool ParseSettingFile(string location = null)
+   {
+      string path = Path.Combine(System.IO.Directory.GetCurrentDirectory(), "spmi_settings.json");
```
Comment (adiaaida) [inline src/collect-all/Utilities.cs]: We want to be careful with this. What if someone is calling your superpmi tool from a different directory than exists spmi_settings.json. You'll at the very least what to confirm that the path exists.

File: src/collect-all/CrossGenTask.cs [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+
+      this.m_command = this.m_command = $"-mch {defaultMchLocation} ";
+      // tempPath should be set on windows to deal with the 255 char limit
+      // in paths.
+      if (instance.m_tempPath != null)
+      {
+         this.m_command += $"-temp {instance.m_tempPath} ";
+      }
+
+      string path = "crossgen_items.json";
```
Comment (adiaaida) [inline src/collect-all/CrossGenTask.cs]: Make sure the file exists. I wonder if this should also be an absolute path?

File: src/superpmi-collect/superpmicollect.cs [hunk 10]
```diff
@@ -0,0 +1,777 @@
...
+            }
+        }
+
+        private static int RunProgram(string program, string arguments)
+        {
+            if (program.LastIndexOf(".py") != -1)
+            {
+                string programArgumentSep = String.IsNullOrEmpty(arguments) ? "" : " ";
+                arguments = "-u " + program; 
+                program = "python";
```
Comment (adiaaida) [inline src/superpmi-collect/superpmicollect.cs]: I'm not sure I like the overwriting of program.

File: src/superpmi-collect/superpmicollect.cs [hunk 11]
```diff
@@ -0,0 +1,777 @@
...
+                s_finalMchFile = Path.GetFullPath(outputMchPath);
+                s_tocFile = s_finalMchFile + ".mct";
+            }
+        }
+
+        private static int RunProgram(string program, string arguments)
+        {
+            if (program.LastIndexOf(".py") != -1)
+            {
+                string programArgumentSep = String.IsNullOrEmpty(arguments) ? "" : " ";
```
Comment (adiaaida) [inline src/superpmi-collect/superpmicollect.cs]: Where is this used?

File: src/superpmi-collect/superpmicollect.cs [hunk 12]
```diff
@@ -0,0 +1,777 @@
...
+                arguments = "-u " + program;
```
Comment (adiaaida) [inline src/superpmi-collect/superpmicollect.cs]: Do we completely drop arguments then? I don't really like the overwritting of arguments

File: src/superpmi-collect/superpmicollect.cs [hunk 13]
```diff
@@ -0,0 +1,777 @@
...
+            
+            // If the program is a script, move the program name into the arguments, and run it
+            // under the appropriate shell.
+            if (Global.IsWindows)
+            {
+                if ((program.LastIndexOf(".bat") != -1) || (program.LastIndexOf(".cmd") != -1))
+                {
+                    string programArgumentSep = String.IsNullOrEmpty(arguments) ? "" : " ";
+                    arguments = "/c " + program + programArgumentSep + arguments;
+                    program = Environment.GetEnvironmentVariable("ComSpec"); // path to CMD.exe
```
Comment (adiaaida) [inline src/superpmi-collect/superpmicollect.cs]: What if we are running in a SysWOW64 window? ComSpec is going to be the normal system32 cmd.exe, not the SysWOW64 cmd.exe. I'm not sure if this makes a difference for superpmi, but it's something to be aware of

File: src/superpmi-collect/superpmicollect.cs [hunk 14]
```diff
@@ -0,0 +1,777 @@
...
+        private static void RunTest(string testName)
+        {
+            string testDir;
+
+            if (Global.IsWindows)
+            {
+                int lastIndex = testName.LastIndexOf("\\");
+                if (lastIndex == -1)
+                {
+                    throw new SpmiException("test path doesn't have any directory separators? " + testName);
```
Comment (adiaaida) [inline src/superpmi-collect/superpmicollect.cs]: Is this basically testing to make sure that we are supplying full paths to the tests? Is that necessary?

File: src/superpmi-collect/superpmicollect.cs [hunk 15]
```diff
@@ -0,0 +1,777 @@
...
+
+                testName = testName.Replace("\\", "/");
+                testName = testName.Replace(".cmd", ".sh");
+                testName = testName.Replace(".bat", ".sh");
+
+                // The way tests are run on Linux, we might need to do some setup. In particular,
+                // if the test scripts are copied from Windows, we need to convert line endings
+                // to Unix line endings, and make the script executable. We can always do this
+                // more than once. This same transformation is done in runtest.sh.
+                // Review: RunProgram doesn't seem to work if the program isn't a full path.
```
Comment (adiaaida) [inline src/superpmi-collect/superpmicollect.cs]: Maybe move the check to make sure we have a absolute path before the RunProgram calls?

File: src/superpmi-collect/superpmicollect.cs [hunk 16]
```diff
@@ -0,0 +1,777 @@
...
+
+                RunProgram("/usr/bin/perl", @"-pi -e 's/\r\n|\n|\r/\n/g' " + "\"" + testName + "\"");
+                RunProgram("/bin/chmod", "+x \"" + testName + "\"");
+
+                // Now, figure out how to run the test.
+
+                int lastIndex = testName.LastIndexOf("/");
+                if (lastIndex == -1)
+                {
+                    throw new SpmiException("test path doesn't have any directory separators? " + testName);
```
Comment (adiaaida) [inline src/superpmi-collect/superpmicollect.cs]: Change the error message to something about absolute paths.

File: src/superpmi-collect/superpmicollect.cs [hunk 17]
```diff
@@ -0,0 +1,777 @@
...
+                @"JIT\Performance\CodeQuality\Roslyn\CscBench\CscBench.cmd",
+                @"JIT\Methodical\fp\exgen\10w5d_cs_do\10w5d_cs_do.cmd",
+                @"JIT\Generics\Coverage\chaos65204782cs\chaos65204782cs.cmd"
+            };
+
+            // Figure out the root of the test binaries directory.
+            // Perhaps this (or something similar) would be a better way to figure out the binary root dir:
+            // testBinaryRootDir = System.IO.Path.GetDirectoryName(new Uri(Assembly.GetExecutingAssembly().CodeBase).LocalPath);
+
+            string thisTestDir = Directory.GetCurrentDirectory();
```
Comment (adiaaida) [inline src/superpmi-collect/superpmicollect.cs]: I'm going to assume at some point we change dirs to a test directory before running this function?

File: src/superpmi-collect/superpmicollect.cs [hunk 18]
```diff
@@ -0,0 +1,777 @@
...
+            int lastIndex = thisTestDir.LastIndexOf("JIT");
+            if (lastIndex == -1)
+            {
+                throw new SpmiException("we expect the current directory when the test is run to be within the JIT test binaries tree, but it is not: " + thisTestDir);
```
Comment (adiaaida) [inline src/superpmi-collect/superpmicollect.cs]: nit: capitalize "we"

File: src/superpmi-collect/superpmicollect.cs [hunk 19]
```diff
@@ -0,0 +1,777 @@
...
+
+        // Merge MC files:
+        //      mcs -merge <s_baseMchFile> <s_tempDir>\*.mc -recursive
+        private static void MergeMCFiles()
+        {
+            string pattern = Path.Combine(s_tempDir, "*.mc");
+            RunProgram(Global.McsPath, "-merge " + s_baseMchFile + " " + pattern + " -recursive");
+            if (!File.Exists(s_baseMchFile))
+            {
+                throw new SpmiException("file missing: " + s_baseMchFile);
```
Comment (adiaaida) [inline src/superpmi-collect/superpmicollect.cs]: nit: capitalization

File: src/superpmi-collect/superpmicollect.cs [hunk 20]
```diff
@@ -0,0 +1,777 @@
...
+                // null out s_baseMchFile so we don't try to delete the same file twice.
+                // Note that we never use s_baseMchFile after this function is called.
+
+                s_cleanMchFile = s_baseMchFile;
+                s_baseMchFile = null;
+            }
+
+            if (!File.Exists(s_cleanMchFile))
+            {
+                throw new SpmiException("file missing: " + s_cleanMchFile);
```
Comment (adiaaida) [inline src/superpmi-collect/superpmicollect.cs]: nit: capitalization.

File: src/superpmi-collect/superpmicollect.cs (hunk 19 above)
Comment (adiaaida) [inline src/superpmi-collect/superpmicollect.cs]: Also, can the exception messages be more verbose as to what likely happened?

File: src/superpmi-collect/superpmicollect.cs (hunk 20 above)
Comment (adiaaida) [inline src/superpmi-collect/superpmicollect.cs]: More verbose error message

File: src/superpmi-collect/superpmicollect.cs [hunk 21]
```diff
@@ -0,0 +1,777 @@
...
+
+        // Create a thin unique MCH:
+        //      <mcl> -removeDup -thin <s_cleanMchFile> <s_finalMchFile>
+        private static void CreateThinUniqueMCH()
+        {
+            RunProgram(Global.McsPath, "-removeDup -thin " + s_cleanMchFile + " " + s_finalMchFile);
+
+            if (!File.Exists(s_finalMchFile))
+            {
+                throw new SpmiException("file missing: " + s_finalMchFile);
```
Comment (adiaaida) [inline src/superpmi-collect/superpmicollect.cs]: nit: capitalization

And a more verbose error message

File: src/superpmi-collect/superpmicollect.cs [hunk 22]
```diff
@@ -0,0 +1,777 @@
...
+        // Create a TOC file:
+        //      <mcl> -toc <s_finalMchFile>
+        //      // check that .mct file was created
+        private static void CreateTOC()
+        {
+            RunProgram(Global.McsPath, "-toc " + s_finalMchFile);
+
+            if (!File.Exists(s_tocFile))
+            {
+                throw new SpmiException("file missing: " + s_tocFile);
```
Comment (adiaaida) [inline src/superpmi-collect/superpmicollect.cs]: nit: capitalization

And a more verbose error message

File: src/superpmi-collect/superpmicollect.cs [hunk 23]
```diff
@@ -0,0 +1,777 @@
...
+        //      <superPmiPath> -p -f <s_finalFailMclFile> <s_finalMchFile> <jitPath>
+        //      if <s_finalFailMclFile> is non-empty:
+        //           // error!
+        private static void VerifyFinalMCH()
+        {
+            RunProgram(Global.SuperPmiPath, "-p -f " + s_finalFailMclFile + " " + s_finalMchFile + " " + Global.JitPath);
+
+            if (!File.Exists(s_finalFailMclFile) || !String.IsNullOrEmpty(File.ReadAllText(s_finalFailMclFile)))
+            {
+                throw new SpmiException("replay of final file is not error free");
```
Comment (adiaaida) [inline src/superpmi-collect/superpmicollect.cs]: nit: capitalization

File: src/superpmi-collect/superpmicollect.cs [hunk 24]
```diff
@@ -0,0 +1,777 @@
...
+                result = 0;
+            }
+            catch (SpmiException ex)
+            {
+                Console.Error.WriteLine("ERROR: " + ex.Message);
+                result = 1;
+            }
+            catch (Exception ex)
+            {
+                Console.Error.WriteLine("ERROR: unknown exception running collection: " + ex.Message);
```
Comment (adiaaida) [inline src/superpmi-collect/superpmicollect.cs]: nit: capitalize unknown

File: src/superpmi-collect/superpmicollect.cs [hunk 25]
```diff
@@ -0,0 +1,777 @@
...
+            }
+
+            return result;
+        }
+    }
+
+    internal class Program
+    {
+        private static void Usage()
+        {
```
Comment (adiaaida) [inline src/superpmi-collect/superpmicollect.cs]: Why not use the same argument utility used by the other tools?

File: src/superpmi-collect/superpmicollect.cs [hunk 26]
```diff
@@ -0,0 +1,777 @@
...
+            // Done with argument parsing.
+
+            string altjitvar = System.Environment.GetEnvironmentVariable("COMPlus_AltJit");
+            if (!String.IsNullOrEmpty(altjitvar))
+            {
+                // Someone already has the COMPlus_AltJit variable set. We don't want to override
+                // that. Perhaps someone is already doing a SuperPMI collection and invokes this
+                // program as part of a full test path in which this program exists.
+
+                Console.WriteLine("COMPlus_AltJit already exists: skipping SuperPMI collection and returning success");
```
Comment (adiaaida) [inline src/superpmi-collect/superpmicollect.cs]: I wouldn't necessarily return success

File: src/superpmi-driver/SuperPMIDriver.cs [hunk 27]
```diff
@@ -0,0 +1,83 @@
...
+            Console.WriteLine("Usage:");
+            Console.WriteLine("SuperPMIDriver.exe <arch> <buildType> <asmDiffs> <coreRootDir>");
+        }
+        
+        public static bool ParseArgs(string[] args)
+        {
+            Options = new Dictionary<string, string>();
+            
+            if (args.Length < 4)
+            {
```
Comment (adiaaida) [inline src/superpmi-driver/SuperPMIDriver.cs]: Why not use the argument utility used in the other tools?

File: src/superpmicrossgen/superpmicrossgen.cs [hunk 28]
```diff
@@ -0,0 +1,326 @@
...
+        private static bool IsAssembly(string filePath)
+        {
+            try
+            {
+                System.Reflection.AssemblyName diffAssembly =
+                    System.Runtime.Loader.AssemblyLoadContext.GetAssemblyName(filePath);
+            }
+            catch (System.IO.FileNotFoundException)
+            {
+                // File not found - not an assembly
```
Comment (adiaaida) [inline src/superpmicrossgen/superpmicrossgen.cs]: Maybe in verbose mode, print out that the file wasn't found?

**Comment by adiaaida** [comment]
Can you also update the doc/getstarted.md file with usage?

File: src/collect-all/Utilities.cs (hunk 1 above)
Comment (jashook) [inline src/collect-all/Utilities.cs]: addressed

File: src/collect-all/Utilities.cs [hunk 29]
```diff
@@ -0,0 +1,479 @@
...
+         CopyDir(subDir.FullName, tempPath);
+      }
+   }
+
+   public void DownloadTests()
+   {
+      Directory.CreateDirectory(this.m_testRoot);
+
+      // Find the last successful test build based on OS.
+      string url = "http://dotnet-ci.cloudapp.net/job/dotnet_coreclr/job/master/job/release_windows_nt/lastSuccessfulBuild/artifact/bin/tests/tests.zip";
```
Comment (russellhadley) [inline src/collect-all/Utilities.cs]: The cijobs tool does some of this today, might be better to not duplicate functionality.

**Comment by BruceForstall** [comment]
@jashook Do you still intend to get back to this? Or, have you transferred ownership to @sandreenko ?
cc @RussKeldorph

**Comment by jashook** [comment]
checked into coreclr master

**Pull request closed without merging**